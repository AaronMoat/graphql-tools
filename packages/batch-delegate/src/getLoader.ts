import { getNamedType, GraphQLOutputType, GraphQLList, GraphQLSchema, FieldNode } from 'graphql';

import DataLoader from 'dataloader';

import {
  SubschemaConfig,
  Transformer,
  createRequestFromInfo,
  getDelegationContext,
  getDelegatingOperation,
  getExecutor,
  validateRequest,
  Receiver,
  externalValueFromResult,
} from '@graphql-tools/delegate';
import { isAsyncIterable, relocatedError } from '@graphql-tools/utils';

import { BatchDelegateOptions } from './types';

const cache1: WeakMap<
  ReadonlyArray<FieldNode>,
  WeakMap<GraphQLSchema | SubschemaConfig, Record<string, DataLoader<any, any>>>
> = new WeakMap();

function createBatchFn<K = any>(options: BatchDelegateOptions) {
  const argsFromKeys = options.argsFromKeys ?? ((keys: ReadonlyArray<K>) => ({ ids: keys }));
  const { lazyOptionsFn } = options;

  return async (keys: ReadonlyArray<K>) => {
    const {
      context,
      info,
      operationName,
      operation = getDelegatingOperation(info.parentType, info.schema),
      fieldName = info.fieldName,
      returnType = new GraphQLList(getNamedType(options.info.returnType) as GraphQLOutputType),
      selectionSet,
      fieldNodes,
      binding,
      skipValidation,
    } = options;

    if (operation !== 'query' && operation !== 'mutation') {
      throw new Error(`Batch delegation not possible for operation '${operation}'.`);
    }

    const request = createRequestFromInfo({
      info,
      operation,
      fieldName,
      selectionSet,
      fieldNodes,
      operationName,
    });

    const delegationContext = getDelegationContext({
      request,
      args: argsFromKeys(keys),
      onLocatedError: originalError => relocatedError(originalError, originalError.path.slice(1)),
      ...(lazyOptionsFn == null ? options : lazyOptionsFn(options)),
      operation,
      fieldName,
      returnType,
    });

    const transformer = new Transformer(delegationContext, binding);

    const processedRequest = transformer.transformRequest(request);

    if (!skipValidation) {
      validateRequest(delegationContext, processedRequest.document);
    }

    const executor = getExecutor(delegationContext);

    const batchResult = await executor({
      ...processedRequest,
      context,
      info,
    });

    if (isAsyncIterable(batchResult)) {
      // TODO: split the asyncIterable and make a new receiver from each of them, return the Receiver instead of the
      // initial value, so that the correct info can be used to instantiate the Receiver
      const receiver = new Receiver(batchResult, delegationContext, executionResult =>
        transformer.transformResult(executionResult)
      );

      const batchValue = await receiver.getInitialValue();

      return Array.isArray(batchValue) ? batchValue : keys.map(() => batchValue);
    }

    // TODO: split the batchedResult and return the result instead of the value, so the correct info
    // can be used to instantiate the value
    const batchValue = externalValueFromResult(transformer.transformResult(batchResult), delegationContext);

    return Array.isArray(batchValue) ? batchValue : keys.map(() => batchValue);
  };
}

export function getLoader<K = any, V = any, C = K>(options: BatchDelegateOptions): DataLoader<K, V, C> {
  const fieldName = options.fieldName ?? options.info.fieldName;

  let cache2: WeakMap<GraphQLSchema | SubschemaConfig, Record<string, DataLoader<K, V, C>>> = cache1.get(
    options.info.fieldNodes
  );

  if (cache2 === undefined) {
    cache2 = new WeakMap();
    cache1.set(options.info.fieldNodes, cache2);
    const loaders = Object.create(null);
    cache2.set(options.schema, loaders);
    const batchFn = createBatchFn(options);
    const loader = new DataLoader<K, V, C>(keys => batchFn(keys), options.dataLoaderOptions);
    loaders[fieldName] = loader;
    return loader;
  }

  let loaders = cache2.get(options.schema);

  if (loaders === undefined) {
    loaders = Object.create(null);
    cache2.set(options.schema, loaders);
    const batchFn = createBatchFn(options);
    const loader = new DataLoader<K, V, C>(keys => batchFn(keys), options.dataLoaderOptions);
    loaders[fieldName] = loader;
    return loader;
  }

  let loader = loaders[fieldName];

  if (loader === undefined) {
    const batchFn = createBatchFn(options);
    loader = new DataLoader<K, V, C>(keys => batchFn(keys), options.dataLoaderOptions);
    loaders[fieldName] = loader;
  }

  return loader;
}
