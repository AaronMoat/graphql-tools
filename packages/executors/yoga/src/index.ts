import { GraphQLSchema } from 'graphql';
import { Plugin } from 'graphql-yoga';
import { useExecutor as useEnvelopExecutor } from '@graphql-tools/executor-envelop';
import { Executor, isPromise, MaybePromise } from '@graphql-tools/utils';
import { schemaFromExecutor } from '@graphql-tools/wrap';

type Opts = Parameters<typeof schemaFromExecutor>[2] & {
  polling?: number;
};

export function useExecutor(
  executor: Executor,
  opts?: Opts,
): Plugin & { invalidateSupergraph: () => void } {
  let schema$: MaybePromise<GraphQLSchema> | undefined;
  let schema: GraphQLSchema | undefined;
  if (opts?.polling) {
    setInterval(() => {
      schema$ = undefined;
      schema = undefined;
    }, opts.polling);
  }
  return {
    onPluginInit({ addPlugin }) {
      addPlugin(
        // @ts-expect-error TODO: fix typings
        useEnvelopExecutor(executor),
      );
    },
    onRequestParse({ serverContext }) {
      return {
        onRequestParseDone() {
          if (!schema$) {
            schema$ ||= schemaFromExecutor(executor, serverContext, opts);
            if (isPromise(schema$)) {
              return schema$.then(newSchema => {
                schema = newSchema;
              }) as Promise<void>;
            }
          }
        },
      };
    },
    onEnveloped({ setSchema }) {
      if (!schema) {
        throw new Error(
          `You provide a promise of a schema but it hasn't been resolved yet. Make sure you use this plugin with GraphQL Yoga.`,
        );
      }
      setSchema(schema);
    },
    invalidateSupergraph() {
      schema$ = undefined;
      schema = undefined;
    },
  };
}
