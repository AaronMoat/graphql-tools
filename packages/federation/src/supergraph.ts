import {
  buildASTSchema,
  DocumentNode,
  EnumTypeDefinitionNode,
  EnumValueDefinitionNode,
  FieldDefinitionNode,
  Kind,
  NamedTypeNode,
  ObjectTypeDefinitionNode,
  OperationTypeNode,
  parse,
  ScalarTypeDefinitionNode,
  TypeDefinitionNode,
  UnionTypeDefinitionNode,
  visit,
} from 'graphql';
import { MergedTypeConfig, SubschemaConfig } from '@graphql-tools/delegate';
import { buildHTTPExecutor } from '@graphql-tools/executor-http';
import { stitchSchemas } from '@graphql-tools/stitch';
import type { Executor } from '@graphql-tools/utils';
import {
  filterInternalFieldsAndTypes,
  getArgsFromKeysForFederation,
  getKeyForFederation,
} from './utils.js';

export interface GetSubschemasFromSupergraphSdlOpts {
  supergraphSdl: string | DocumentNode;
  onExecutor?: (opts: { subgraphName: string; endpoint: string }) => Executor;
  batch?: boolean;
}

export function getSubschemasFromSupergraphSdl({
  supergraphSdl,
  onExecutor = ({ endpoint }) => buildHTTPExecutor({ endpoint }),
  batch = false,
}: GetSubschemasFromSupergraphSdlOpts) {
  const ast =
    typeof supergraphSdl === 'string' ? parse(supergraphSdl, { noLocation: true }) : supergraphSdl;
  const subgraphRootFieldDefinitionNodes = new Map<
    string,
    Map<OperationTypeNode, FieldDefinitionNode[]>
  >();
  const subgraphEndpointMap = new Map<string, string>();
  const subgraphTypesMap = new Map<string, TypeDefinitionNode[]>();
  const typeNameKeyBySubgraphMap = new Map<string, Map<string, string>>();
  const typeNameFieldsKeyBySubgraphMap = new Map<string, Map<string, Map<string, string>>>();
  const typeNameCanonicalMap = new Map<string, string>();
  const rootTypeNames = new Map<OperationTypeNode, string>();
  visit(ast, {
    SchemaDefinition(node) {
      node.operationTypes?.forEach(operationTypeNode => {
        rootTypeNames.set(operationTypeNode.operation, operationTypeNode.type.name.value);
      });
    },
    ScalarTypeDefinition(node) {
      node.directives?.forEach(directiveNode => {
        if (directiveNode.name.value === 'join__type') {
          const graphArgumentNode = directiveNode.arguments?.find(
            argumentNode => argumentNode.name.value === 'graph',
          );
          if (graphArgumentNode?.value?.kind === Kind.ENUM) {
            const graphName = graphArgumentNode.value.value;
            let subgraphTypes = subgraphTypesMap.get(graphName);
            if (!subgraphTypes) {
              subgraphTypes = [];
              subgraphTypesMap.set(graphName, subgraphTypes);
            }
            subgraphTypes.push({
              ...node,
              directives: node.directives?.filter(
                directiveNode => directiveNode.name.value !== 'join__type',
              ),
            });
          }
        }
      });
    },
    InputObjectTypeDefinition(node) {
      node.directives?.forEach(directiveNode => {
        if (directiveNode.name.value === 'join__type') {
          const graphArgumentNode = directiveNode.arguments?.find(
            argumentNode => argumentNode.name.value === 'graph',
          );
          if (graphArgumentNode?.value?.kind === Kind.ENUM) {
            const graphName = graphArgumentNode.value.value;
            let subgraphTypes = subgraphTypesMap.get(graphName);
            if (!subgraphTypes) {
              subgraphTypes = [];
              subgraphTypesMap.set(graphName, subgraphTypes);
            }
            subgraphTypes.push({
              ...node,
              directives: node.directives?.filter(
                directiveNode => directiveNode.name.value !== 'join__type',
              ),
            });
          }
        }
      });
    },
    InterfaceTypeDefinition(node) {
      node.directives?.forEach(directiveNode => {
        if (directiveNode.name.value === 'join__type') {
          const graphArgumentNode = directiveNode.arguments?.find(
            argumentNode => argumentNode.name.value === 'graph',
          );
          if (graphArgumentNode?.value?.kind === Kind.ENUM) {
            const graphName = graphArgumentNode.value.value;
            let subgraphTypes = subgraphTypesMap.get(graphName);
            if (!subgraphTypes) {
              subgraphTypes = [];
              subgraphTypesMap.set(graphName, subgraphTypes);
            }
            subgraphTypes.push({
              ...node,
              directives: node.directives?.filter(
                directiveNode => directiveNode.name.value !== 'join__type',
              ),
            });
          }
        }
      });
    },
    UnionTypeDefinition(node) {
      node.directives?.forEach(directiveNode => {
        if (directiveNode.name.value === 'join__type') {
          const graphArgumentNode = directiveNode.arguments?.find(
            argumentNode => argumentNode.name.value === 'graph',
          );
          if (graphArgumentNode?.value?.kind === Kind.ENUM) {
            const graphName = graphArgumentNode.value.value;
            const members = node.directives?.filter(
              directiveNode =>
                directiveNode.name.value === 'join__unionMember' &&
                directiveNode.arguments?.find(
                  argumentNode =>
                    argumentNode.name.value === 'graph' &&
                    argumentNode.value?.kind === Kind.ENUM &&
                    argumentNode.value.value === graphName,
                ),
            );
            if (members?.length) {
              let subgraphTypes = subgraphTypesMap.get(graphName);
              if (!subgraphTypes) {
                subgraphTypes = [];
                subgraphTypesMap.set(graphName, subgraphTypes);
              }
              subgraphTypes.push({
                ...node,
                directives: node.directives?.filter(
                  directiveNode =>
                    directiveNode.name.value !== 'join__type' &&
                    directiveNode.name.value !== 'join__unionMember',
                ),
              });
            }
          }
        }
      });
    },
    EnumTypeDefinition(node) {
      if (node.name.value === 'join__Graph') {
        node.values?.forEach(valueNode => {
          const joinGraphDirectiveNode = valueNode.directives?.find(
            directiveNode => directiveNode.name.value === 'join__graph',
          );
          if (joinGraphDirectiveNode) {
            const urlArgumentNode = joinGraphDirectiveNode.arguments?.find(
              argumentNode => argumentNode.name.value === 'url',
            );
            if (urlArgumentNode?.value?.kind === Kind.STRING) {
              subgraphEndpointMap.set(valueNode.name.value, urlArgumentNode.value.value);
            }
          }
        });
      }
      node.directives?.forEach(directiveNode => {
        if (directiveNode.name.value === 'join__type') {
          const graphArgumentNode = directiveNode.arguments?.find(
            argumentNode => argumentNode.name.value === 'graph',
          );
          if (graphArgumentNode?.value?.kind === Kind.ENUM) {
            const graphName = graphArgumentNode.value.value;
            const enumValueNodes: EnumValueDefinitionNode[] = [];
            node.values?.forEach(valueNode => {
              const joinEnumValueDirectiveNode = valueNode.directives?.find(
                directiveNode => directiveNode.name.value === 'join__enumValue',
              );
              if (joinEnumValueDirectiveNode) {
                const graphArgumentNode = joinEnumValueDirectiveNode.arguments?.find(
                  argumentNode => argumentNode.name.value === 'graph',
                );
                if (graphArgumentNode?.value?.kind === Kind.ENUM) {
                  if (graphArgumentNode.value.value === graphName) {
                    enumValueNodes.push({
                      ...valueNode,
                      directives: valueNode.directives?.filter(
                        directiveNode => directiveNode.name.value !== 'join__enumValue',
                      ),
                    });
                  }
                }
              }
            });

            const enumTypedDefNodeForSubgraph: EnumTypeDefinitionNode = {
              ...node,
              directives: node.directives?.filter(
                directiveNode => directiveNode.name.value !== 'join__type',
              ),
              values: enumValueNodes,
            };
            let subgraphTypes = subgraphTypesMap.get(graphName);
            if (!subgraphTypes) {
              subgraphTypes = [];
              subgraphTypesMap.set(graphName, subgraphTypes);
            }
            subgraphTypes.push(enumTypedDefNodeForSubgraph);
          }
        }
      });
    },
    ObjectTypeDefinition(node) {
      if ([...rootTypeNames.values()].includes(node.name.value)) {
        const operationTypeName = [...rootTypeNames.entries()].find(
          ([, rootTypeName]) => rootTypeName === node.name.value,
        )![0];
        node.fields?.forEach(fieldNode => {
          fieldNode.directives?.forEach(directiveNode => {
            if (directiveNode.name.value === 'join__field') {
              const graphArgumentNode = directiveNode.arguments?.find(
                argumentNode => argumentNode.name.value === 'graph',
              );
              if (graphArgumentNode?.value?.kind === Kind.ENUM) {
                const graphName = graphArgumentNode.value.value;
                let subgraphRootFieldDefinitionNodeMap =
                  subgraphRootFieldDefinitionNodes.get(graphName);
                if (!subgraphRootFieldDefinitionNodeMap) {
                  subgraphRootFieldDefinitionNodeMap = new Map();
                  subgraphRootFieldDefinitionNodes.set(
                    graphName,
                    subgraphRootFieldDefinitionNodeMap,
                  );
                }
                let fieldDefinitionNodesOfSubgraph =
                  subgraphRootFieldDefinitionNodeMap.get(operationTypeName);
                if (!fieldDefinitionNodesOfSubgraph) {
                  fieldDefinitionNodesOfSubgraph = [];
                  subgraphRootFieldDefinitionNodeMap.set(
                    operationTypeName,
                    fieldDefinitionNodesOfSubgraph,
                  );
                }
                fieldDefinitionNodesOfSubgraph.push({
                  ...fieldNode,
                  directives: fieldNode.directives?.filter(
                    directiveNode => directiveNode.name.value !== 'join__field',
                  ),
                });
              }
            }
          });
        });
      }
      node.directives?.forEach(directiveNode => {
        if (directiveNode.name.value === 'join__owner') {
          const graphArgumentNode = directiveNode.arguments?.find(
            argumentNode => argumentNode.name.value === 'graph',
          );
          if (graphArgumentNode?.value?.kind === Kind.ENUM) {
            typeNameCanonicalMap.set(node.name.value, graphArgumentNode.value.value);
          }
        }
        if (directiveNode.name.value === 'join__type') {
          const graphArgumentNode = directiveNode.arguments?.find(
            argumentNode => argumentNode.name.value === 'graph',
          );
          if (graphArgumentNode?.value?.kind === Kind.ENUM) {
            let keyFieldName: string;
            const graphName = graphArgumentNode.value.value;
            const keyArgumentNode = directiveNode.arguments?.find(
              argumentNode => argumentNode.name.value === 'key',
            );
            if (keyArgumentNode?.value?.kind === Kind.STRING) {
              let typeNameKeyMap = typeNameKeyBySubgraphMap.get(graphName);
              if (!typeNameKeyMap) {
                typeNameKeyMap = new Map();
                typeNameKeyBySubgraphMap.set(graphName, typeNameKeyMap);
              }
              keyFieldName = keyArgumentNode.value.value;
              typeNameKeyMap.set(node.name.value, keyFieldName);
            }
            const fieldDefinitionNodesOfSubgraph: FieldDefinitionNode[] = [];
            node.fields?.forEach(fieldNode => {
              if (fieldNode.name.value === keyFieldName) {
                fieldDefinitionNodesOfSubgraph.push({
                  ...fieldNode,
                  directives: fieldNode.directives?.filter(
                    directiveNode => directiveNode.name.value !== 'join__field',
                  ),
                });
                return;
              }
              const joinFieldDirectiveNode = fieldNode.directives?.find(
                directiveNode => directiveNode.name.value === 'join__field',
              );
              if (joinFieldDirectiveNode) {
                const graphArgumentNode = joinFieldDirectiveNode.arguments?.find(
                  argumentNode => argumentNode.name.value === 'graph',
                );
                if (graphArgumentNode?.value?.kind === Kind.ENUM) {
                  if (graphArgumentNode.value.value === graphName) {
                    const requiresArgumentNode = joinFieldDirectiveNode.arguments?.find(
                      argumentNode => argumentNode.name.value === 'requires',
                    );
                    if (requiresArgumentNode?.value?.kind === Kind.STRING) {
                      let typeNameFieldsKeyMap = typeNameFieldsKeyBySubgraphMap.get(graphName);
                      if (!typeNameFieldsKeyMap) {
                        typeNameFieldsKeyMap = new Map();
                        typeNameFieldsKeyBySubgraphMap.set(graphName, typeNameFieldsKeyMap);
                      }
                      let fieldsKeyMap = typeNameFieldsKeyMap.get(node.name.value);
                      if (!fieldsKeyMap) {
                        fieldsKeyMap = new Map();
                        typeNameFieldsKeyMap.set(node.name.value, fieldsKeyMap);
                      }
                      fieldsKeyMap.set(fieldNode.name.value, requiresArgumentNode.value.value);
                    }
                    fieldDefinitionNodesOfSubgraph.push({
                      ...fieldNode,
                      directives: fieldNode.directives?.filter(
                        directiveNode => directiveNode.name.value !== 'join__field',
                      ),
                    });
                  }
                }
              } else {
                fieldDefinitionNodesOfSubgraph.push({
                  ...fieldNode,
                  directives: fieldNode.directives?.filter(
                    directiveNode => directiveNode.name.value !== 'join__field',
                  ),
                });
              }
            });
            const objectTypedDefNodeForSubgraph: ObjectTypeDefinitionNode = {
              ...node,
              fields: fieldDefinitionNodesOfSubgraph,
              directives: node.directives?.filter(
                directiveNode =>
                  directiveNode.name.value !== 'join__type' &&
                  directiveNode.name.value !== 'join__owner',
              ),
            };
            let subgraphTypes = subgraphTypesMap.get(graphName);
            if (!subgraphTypes) {
              subgraphTypes = [];
              subgraphTypesMap.set(graphName, subgraphTypes);
            }
            subgraphTypes.push(objectTypedDefNodeForSubgraph);
          }
        }
      });
    },
  });
  const subschemaMap = new Map<string, SubschemaConfig>();
  for (const [subgraphName, endpoint] of subgraphEndpointMap) {
    const executor = onExecutor({ subgraphName, endpoint });
    const mergeConfig: SubschemaConfig['merge'] = {};
    const typeNameKeyMap = typeNameKeyBySubgraphMap.get(subgraphName);
    const unionTypeNodes: NamedTypeNode[] = [];
    if (typeNameKeyMap) {
      const typeNameFieldsKeyMap = typeNameFieldsKeyBySubgraphMap.get(subgraphName);
      for (const [typeName, key] of typeNameKeyMap) {
        const mergedTypeConfig: MergedTypeConfig = (mergeConfig[typeName] = {
          selectionSet: `{ ${key} }`,
          argsFromKeys: getArgsFromKeysForFederation,
          key: getKeyForFederation,
          fieldName: `_entities`,
        });
        const keyProps = key.split(' ');
        mergedTypeConfig.dataLoaderOptions = {
          cacheKeyFn(root) {
            return keyProps.map(key => root[key]).join(' ');
          },
        };
        const fieldsKeyMap = typeNameFieldsKeyMap?.get(typeName);
        if (fieldsKeyMap) {
          const fieldsConfig = (mergedTypeConfig.fields = {});
          for (const [fieldName, key] of fieldsKeyMap) {
            fieldsConfig[fieldName] = {
              selectionSet: `{ ${key} }`,
              computed: true,
            };
          }
        }
        if (typeNameCanonicalMap.get(typeName) === subgraphName) {
          mergedTypeConfig.canonical = true;
        }
        unionTypeNodes.push({
          kind: Kind.NAMED_TYPE,
          name: {
            kind: Kind.NAME,
            value: typeName,
          },
        });
      }
    }
    const entitiesUnionTypeDefinitionNode: UnionTypeDefinitionNode = {
      name: {
        kind: Kind.NAME,
        value: '_Entity',
      },
      kind: Kind.UNION_TYPE_DEFINITION,
      types: unionTypeNodes,
    };
    let subgraphRootFieldDefinitionNodeMap = subgraphRootFieldDefinitionNodes.get(subgraphName);
    if (!subgraphRootFieldDefinitionNodeMap) {
      subgraphRootFieldDefinitionNodeMap = new Map();
      subgraphRootFieldDefinitionNodes.set(subgraphName, subgraphRootFieldDefinitionNodeMap);
    }
    let queryFields = subgraphRootFieldDefinitionNodeMap.get('query' as OperationTypeNode);
    if (!queryFields) {
      queryFields = [];
      subgraphRootFieldDefinitionNodeMap.set('query' as OperationTypeNode, queryFields);
    }
    queryFields.push({
      kind: Kind.FIELD_DEFINITION,
      name: {
        kind: Kind.NAME,
        value: '_entities',
      },
      type: {
        kind: Kind.NAMED_TYPE,
        name: {
          kind: Kind.NAME,
          value: '_Entity',
        },
      },
      arguments: [
        {
          kind: Kind.INPUT_VALUE_DEFINITION,
          name: {
            kind: Kind.NAME,
            value: 'representations',
          },
          type: {
            kind: Kind.NON_NULL_TYPE,
            type: {
              kind: Kind.LIST_TYPE,
              type: {
                kind: Kind.NON_NULL_TYPE,
                type: {
                  kind: Kind.NAMED_TYPE,
                  name: {
                    kind: Kind.NAME,
                    value: '_Any',
                  },
                },
              },
            },
          },
        },
      ],
    });
    const rootTypes: TypeDefinitionNode[] = [];
    for (const [operationType, fieldDefinitionNodes] of subgraphRootFieldDefinitionNodeMap) {
      rootTypes.push({
        kind: Kind.OBJECT_TYPE_DEFINITION,
        name: {
          kind: Kind.NAME,
          value: rootTypeNames.get(operationType)!,
        },
        fields: fieldDefinitionNodes,
      });
    }
    const subgraphTypes = subgraphTypesMap.get(subgraphName) || [];
    const schema = buildASTSchema(
      {
        kind: Kind.DOCUMENT,
        definitions: [
          ...subgraphTypes,
          entitiesUnionTypeDefinitionNode,
          anyTypeDefinitionNode,
          ...rootTypes,
        ],
      },
      {
        assumeValidSDL: true,
        assumeValid: true,
      },
    );
    subschemaMap.set(subgraphName, {
      schema,
      executor,
      merge: mergeConfig,
      batch,
    });
  }
  return subschemaMap;
}

export function getStitchedSchemaFromSupergraphSdl(opts: GetSubschemasFromSupergraphSdlOpts) {
  const subschemaMap = getSubschemasFromSupergraphSdl(opts);
  const supergraphSchema = stitchSchemas({
    subschemas: [...subschemaMap.values()],
    assumeValid: true,
    assumeValidSDL: true,
  });
  return filterInternalFieldsAndTypes(supergraphSchema);
}

const anyTypeDefinitionNode: ScalarTypeDefinitionNode = {
  name: {
    kind: Kind.NAME,
    value: '_Any',
  },
  kind: Kind.SCALAR_TYPE_DEFINITION,
};
