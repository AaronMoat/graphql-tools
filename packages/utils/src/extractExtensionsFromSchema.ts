import { GraphQLFieldConfig, GraphQLSchema } from 'graphql';
import { asArray } from './helpers.js';
import { MapperKind } from './Interfaces.js';
import { mapSchema } from './mapSchema.js';
import {
  EnumTypeExtensions,
  InputTypeExtensions,
  ObjectTypeExtensions,
  SchemaExtensions,
} from './types.js';

function handleDirectiveExtensions(extensions: any, removeDirectives: boolean) {
  extensions = extensions || {};
  const { directives: existingDirectives, ...rest } = extensions;
  const finalExtensions: any = {
    ...rest,
  };
  if (!removeDirectives) {
    if (existingDirectives != null) {
      const directives = {};
      for (const directiveName in existingDirectives) {
        directives[directiveName] = [...asArray(existingDirectives[directiveName])];
      }
      finalExtensions.directives = directives;
    }
  }
  return finalExtensions;
}

export function extractExtensionsFromSchema(
  schema: GraphQLSchema,
  removeDirectives = false,
): SchemaExtensions {
  const result: SchemaExtensions = {
    schemaExtensions: handleDirectiveExtensions(schema.extensions, removeDirectives),
    types: {},
  };

  mapSchema(schema, {
    [MapperKind.OBJECT_TYPE]: type => {
      result.types[type.name] = {
        fields: {},
        type: 'object',
        extensions: handleDirectiveExtensions(type.extensions, removeDirectives),
      };
      return type;
    },
    [MapperKind.INTERFACE_TYPE]: type => {
      result.types[type.name] = {
        fields: {},
        type: 'interface',
        extensions: handleDirectiveExtensions(type.extensions, removeDirectives),
      };
      return type;
    },
    [MapperKind.FIELD]: (field, fieldName, typeName) => {
      (result.types[typeName] as ObjectTypeExtensions).fields[fieldName] = {
        arguments: {},
        extensions: handleDirectiveExtensions(field.extensions, removeDirectives),
      };
      const args = (field as GraphQLFieldConfig<any, any>).args;
      if (args != null) {
        for (const argName in args) {
          (result.types[typeName] as ObjectTypeExtensions).fields[fieldName].arguments[argName] =
            handleDirectiveExtensions(args[argName].extensions, removeDirectives);
        }
      }
      return field;
    },
    [MapperKind.ENUM_TYPE]: type => {
      result.types[type.name] = {
        values: {},
        type: 'enum',
        extensions: handleDirectiveExtensions(type.extensions, removeDirectives),
      };
      return type;
    },
    [MapperKind.ENUM_VALUE]: (value, typeName, _schema, valueName) => {
      (result.types[typeName] as EnumTypeExtensions).values[valueName] = handleDirectiveExtensions(
        value.extensions,
        removeDirectives,
      );
      return value;
    },
    [MapperKind.SCALAR_TYPE]: type => {
      result.types[type.name] = {
        type: 'scalar',
        extensions: handleDirectiveExtensions(type.extensions, removeDirectives),
      };
      return type;
    },
    [MapperKind.UNION_TYPE]: type => {
      result.types[type.name] = {
        type: 'union',
        extensions: handleDirectiveExtensions(type.extensions, removeDirectives),
      };
      return type;
    },
    [MapperKind.INPUT_OBJECT_TYPE]: type => {
      result.types[type.name] = {
        fields: {},
        type: 'input',
        extensions: handleDirectiveExtensions(type.extensions, removeDirectives),
      };
      return type;
    },
    [MapperKind.INPUT_OBJECT_FIELD]: (field, fieldName, typeName) => {
      (result.types[typeName] as InputTypeExtensions).fields[fieldName] = {
        extensions: handleDirectiveExtensions(field.extensions, removeDirectives),
      };
      return field;
    },
  });

  return result;
}
