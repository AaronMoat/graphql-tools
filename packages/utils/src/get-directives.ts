import {
  GraphQLEnumTypeConfig,
  GraphQLEnumValue,
  GraphQLEnumValueConfig,
  GraphQLField,
  GraphQLFieldConfig,
  GraphQLInputField,
  GraphQLInputFieldConfig,
  GraphQLInputObjectTypeConfig,
  GraphQLInterfaceTypeConfig,
  GraphQLNamedType,
  GraphQLObjectTypeConfig,
  GraphQLScalarTypeConfig,
  GraphQLSchema,
  GraphQLSchemaConfig,
  GraphQLUnionTypeConfig,
} from 'graphql';
import { getDirectiveExtensions } from './getDirectiveExtensions.js';

export interface DirectiveAnnotation {
  name: string;
  args?: Record<string, any>;
}

export type DirectableGraphQLObject =
  | GraphQLSchema
  | GraphQLSchemaConfig
  | GraphQLNamedType
  | GraphQLObjectTypeConfig<any, any>
  | GraphQLInterfaceTypeConfig<any, any>
  | GraphQLUnionTypeConfig<any, any>
  | GraphQLScalarTypeConfig<any, any>
  | GraphQLEnumTypeConfig
  | GraphQLEnumValue
  | GraphQLEnumValueConfig
  | GraphQLInputObjectTypeConfig
  | GraphQLField<any, any>
  | GraphQLInputField
  | GraphQLFieldConfig<any, any>
  | GraphQLInputFieldConfig;

export function getDirectivesInExtensions(
  node: DirectableGraphQLObject,
  pathToDirectivesInExtensions = ['directives'],
): Array<DirectiveAnnotation> {
  const directiveExtensions = getDirectiveExtensions(node, undefined, pathToDirectivesInExtensions);
  return Object.entries(directiveExtensions)
    .map(([directiveName, directiveArgsArr]) =>
      directiveArgsArr?.map(directiveArgs => ({
        name: directiveName,
        args: directiveArgs,
      })),
    )
    .flat(Infinity)
    .filter(Boolean) as Array<DirectiveAnnotation>;
}

export function getDirectiveInExtensions(
  node: DirectableGraphQLObject,
  directiveName: string,
  pathToDirectivesInExtensions = ['directives'],
): Array<Record<string, any>> | undefined {
  const directiveExtensions = getDirectiveExtensions(node, undefined, pathToDirectivesInExtensions);
  return directiveExtensions[directiveName];
}

export function getDirectives(
  schema: GraphQLSchema,
  node: DirectableGraphQLObject,
  pathToDirectivesInExtensions = ['directives'],
): Array<DirectiveAnnotation> {
  const directiveExtensions = getDirectiveExtensions(node, schema, pathToDirectivesInExtensions);
  return Object.entries(directiveExtensions)
    .map(([directiveName, directiveArgsArr]) =>
      directiveArgsArr?.map(directiveArgs => ({
        name: directiveName,
        args: directiveArgs,
      })),
    )
    .flat(Infinity)
    .filter(Boolean) as Array<DirectiveAnnotation>;
}

export function getDirective(
  schema: GraphQLSchema,
  node: DirectableGraphQLObject,
  directiveName: string,
  pathToDirectivesInExtensions = ['directives'],
): Array<Record<string, any>> | undefined {
  const directiveExtensions = getDirectiveExtensions(node, schema, pathToDirectivesInExtensions);
  return directiveExtensions[directiveName];
}
