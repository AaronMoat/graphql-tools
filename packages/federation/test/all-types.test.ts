import { promises as fs } from 'fs';
import { join } from 'path';
import { printSchemaWithDirectives } from '@graphql-tools/utils';
import { getStitchedSchemaFromSupergraphSdl } from '../src/supergraph';

describe('All types', () => {
  it('works', async () => {
    const supergraphSdl = await fs.readFile(
      join(__dirname, 'fixtures/all-types/supergraph.graphql'),
      'utf-8',
    );
    const stitchedSchema = getStitchedSchemaFromSupergraphSdl({
      supergraphSdl,
    });
    const printedSchema = printSchemaWithDirectives(stitchedSchema);
    expect(printedSchema.trim()).toMatchSnapshot();
  });
});
