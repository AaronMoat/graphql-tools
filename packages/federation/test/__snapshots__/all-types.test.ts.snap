// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All types works 1`] = `
"schema {
  query: Query
}

type Query {
  me: User
  users: [User]
  topProducts(first: Int): [Product]
}

type User {
  numberOfReviews: Int
  reviews: [Review]
  username: String
  birthDate: String
  id: ID!
  name: String
}

type Product {
  reviews(input: ProductReviewsInput): [Review]
  inStock: Boolean
  shippingEstimate: Int
  name: String
  price: Int
  upc: String!
  weight: Int
}

enum ReviewsOrder {
  ASC
  DESC
}

input ProductReviewsInput {
  first: Int
  order: ReviewsOrder
}

type AnonymousReview implements Review {
  body: String
  id: ID!
  product: Product
}

interface Review {
  body: String
  id: ID!
  product: Product
}

scalar Stars

type UserReview implements Review {
  author: User
  body: String
  id: ID!
  product: Product
  stars: Stars
}"
`;
